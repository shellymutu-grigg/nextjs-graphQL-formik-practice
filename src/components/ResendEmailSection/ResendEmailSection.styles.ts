import { makeStyles } from '@material-ui/core/styles';

export const useResendEmailStyles = makeStyles((theme) => ({
  headerShadow: {
    boxShadow: '0 4px 4px 0 rgba(12,0,51,0.1)',
  },
  layout: {
    overflow: 'hidden',
    minWidth: '320px',
    height: '100%',
    '& .main-content': {
      position: 'absolute',
      top: '4.5rem',
      margin: '0 auto',
      width: '100%',
      left: '0',
      right: '0',
      padding: '1rem 1rem 6rem',
    },
  },
  boldHeader: {
    fontSize: '1.25rem',
    fontWeight: 'bold',
    lineHeight: '1.75rem',
    paddingTop: '2rem',
  },
  bottomMargin1point5: {
    marginBottom: theme.spacing(1.5),
  },
  padding: {
    paddingTop: '2rem',
  },
}));
