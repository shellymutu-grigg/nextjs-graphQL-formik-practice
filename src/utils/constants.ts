/**
 * Date Formats based on date-fns accepted patterns.
 *
 * @see https://date-fns.org/v2.23.0/docs/format#description
 */
export const dateFormats = {
  dd_MMM_yyyy: 'dd MMM yyyy',
  yyyy_MM_dd: 'yyyy-MM-dd',
  MMMM_yyyy: 'MMMM yyyy',
  dd_MMMM_yyyy: 'dd MMMM yyyy',
  DATE_DISPLAY_FORMAT: 'dd/MM/yyyy',
  DATE_VALUE_FORMAT: 'yyyy-MM-dd',
} as const;

/**
 * Defaults
 */
export const defaults = {
  EMAIL_ADDRESS_MAX_LENGTH: 64,
  FAILURE_STATUSCODE: 500,
  FIRST_NAME_MAX_LENGTH: 50,
  LAST_NAME_MAX_LENGTH: 50,
  RESEND_EMAIL_MAX_LIMIT: 5,
  RESEND_EMAIL_TIME_OUT: 3000,
  SUCCESS_STATUSCODE: 200,
} as const;

/**
 * Defaults
 */
export const errors = {
  NO_MATCHING_DETAILS:
    "Sorry, we couldn't match those details to a Membership.",
  CORRECTLY_ENTER_DETAILS: 'Please check your details are entered correctly.',
  CONTINUED_ERROR_PART_1:
    'If you continue to get this error, please give us a call on',
} as const;

/**
 * Reusable messages used in application
 * @type {Object.<string, string>}
 */
export const messages = {
  CHECK_EMAIL_PART_ONE: "Check the email we've sent to ",
  CHECK_EMAIL_PART_TWO: ' and click the link to verify your email',
  CONTACT_CENTER_EMAIL: 'test@email.com',
  CONTACT_CENTER_PHONE_NUMBER: '0800 GET HELP',
  CONTACT_CENTER_PHONE_NUMBER_TEL_LINK: '0800022022',
  DATE_OF_BIRTH: 'Date of birth',
  EMAIL_SENT_CTA: 'Email Sent',
  ENTER_DETAILS_PART_ONE: "Populate the following fields and click '",
  ENTER_DETAILS_PART_TWO: "' to identify your membership record.",
  FIRST_NAME: 'First name',
  HELP_TEXT_PART_ONE: 'If you need help, ',
  LAST_NAME: 'Last name',
  MESSAGE_TEXT_ONE: 'please email us at ',
  MESSAGE_TEXT_TWO: ' or give us a call on ',
  NEXT_CTA_TEXT: 'Next',
  PHONE_NUMBER: 'Phone number',
  RESEND_CTA: 'Resend email',
  RESEND_EMAIL_ERROR:
    'We seem to be having issues right now. Please try again later.',
  SPAM_FOLDER_CHECK:
    "Remember to check spam and other folders. If you haven't received an email from us, you can " +
    'click the button below in 60 seconds to resend it.',
  VERIFICATION: 'Verification',
} as const;

export const LOGIN_URL = '/api/auth/login';
export const NEXT_PUBLIC_API = '/api/graphql';

/**
 * Slugs for all the pages in the app. This should match the slug paths
 * configured in Contentful.
 * @type {Object.<string, string>} slugPaths
 */
export const slugPaths = {
  RESEND_EMAIL: '/resend-email',
} as const;

/**
 * Field validation messages.
 */
export const validationMessages = {
  DOB: {
    REQUIRED: 'Date of birth is required',
    INVALID: 'Must be a valid date',
    MIN_DATE: 'Must not be more than 100 years in the past',
    MAX_DATE: "Must be earlier than today's date",
  },
  EMAIL_ADDRESS: {
    REQUIRED: 'Email address is required',
    INVALID: 'Must be a valid email address',
  },
  NAME: {
    REQUIRED: 'This field is required',
    ALLOWED_CHARACTERS: 'Must contain alpha and accented characters only',
    MORE_THAN_TWO_SPACES: 'Must not contain more than two space characters',
    NO_LEADING_SPACE: 'Must not start with a space character',
    MAX_LENGTH: 'Must be less than 30 characters',
  },
  MEMBERSHIP_NUMBER: {
    REQUIRED: 'Membership number is required',
    DEFAULT_START_VALUE: 'Must start with 3083 26',
    MIN_MAX_LENGTH: 'Must be 16 digits long',
  },
  PHONE_NUMBER: {
    REQUIRED: 'Phone number is required',
    ALLOWED_CHARACTERS: 'Must start with 020-029 or 03-09',
    MIN_LENGTH: 'Must have at least 9 digits',
    MAX_LENGTH: 'Must be less than 13 digits long',
    AT_LEAST_ONE_NUMBER_IS_REQUIRED:
      'At least one phone number should be populated',
  },
} as const;
