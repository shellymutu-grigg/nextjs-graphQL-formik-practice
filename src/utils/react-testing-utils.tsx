import React from 'react';
import { UserProvider } from '@auth0/nextjs-auth0';
import { render, queries, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from '@material-ui/core/styles';
import AppContextProvider from '@utils/app-context';
import theme from '@theme/theme';
import { ReactQueryClientProvider } from '@components/ReactQueryClientProvider/ReactQueryClientProvider';

const Provider: React.FC = ({ children }) => (
  <ReactQueryClientProvider>
    <AppContextProvider>
      <ThemeProvider theme={theme}>
        <UserProvider>{children}</UserProvider>
      </ThemeProvider>
    </AppContextProvider>
  </ReactQueryClientProvider>
);

const themeRenderer = (
  ui: React.ReactElement,
  options: RenderOptions<typeof queries, HTMLElement> | undefined = {},
) =>
  render(ui, {
    wrapper: Provider,
    ...options,
  });

export { themeRenderer as render };

export * from '@testing-library/react';
