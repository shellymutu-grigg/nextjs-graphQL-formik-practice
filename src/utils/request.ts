import fetch from 'cross-fetch';

const getResponse = <ResponseType = any>({
  response,
  data,
}: {
  response: Response;
  data: ResponseType;
}) => {
  const returnValue = { response, data };

  if (!response.ok) {
    return Promise.reject(returnValue);
  }

  return returnValue;
};

export const request = async <ResponseType = any>(
  url: string,
  requestObj: Omit<RequestInit, 'body'> & { body?: any },
) => {
  if (requestObj && typeof requestObj.body === 'object') {
    requestObj.body = JSON.stringify(requestObj.body);
  }

  const headers = {
    'Content-Type': 'application/json',
    ...requestObj.headers,
  };

  const response = await fetch(url, {
    ...requestObj,
    headers: {
      'Content-Type': 'application/json',
      ...requestObj.headers,
    },
  });

  // HTTP 204 means no content
  if (response.status === 204) {
    return { response, data: null };
  }

  const data =
    headers['Content-Type'] === 'application/json'
      ? ((await response.json()) as ResponseType)
      : await response.text();

  return getResponse({ response, data });
};

const requestWithoutCustomHeader = async <ResponseType = any>(
  url: string,
  requestObj: RequestInit,
) => {
  const response = await fetch(url, requestObj);
  const data = (await response.json()) as ResponseType;
  return getResponse({ response, data });
};

export const requestGet = async (url: string, requestObj: RequestInit) =>
  requestWithoutCustomHeader(url, requestObj);

export const requestFormUrlEncoded = async (
  url: string,
  requestObj: RequestInit,
) => requestWithoutCustomHeader(url, requestObj);
