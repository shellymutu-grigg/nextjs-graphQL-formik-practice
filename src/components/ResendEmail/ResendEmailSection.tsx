import React, { FC } from 'react';
import { Typography, Grid, Box } from '@material-ui/core';
import { messages } from '@utils/constants';
import { obfuscateEmail } from '@utils/obfuscate-email';
import { Button } from '@components/Button/Button';
import { Link } from '@components/Link/Link';
import { useResendEmailStyles } from './ResendEmailSection.styles';

interface CardProps {
  user: any;
  handleClick: () => void;
  isDisabled: boolean;
  apiError: boolean;
}
export const ResendEmailSection: FC<CardProps> = ({
  user,
  handleClick,
  isDisabled,
  apiError,
}) => {
  const { boldHeader, marginBottomX1point5 } = useResendEmailStyles();

  return (
    <>
      <Grid
        component={Grid}
        container
        direction="row"
        justifyContent="center"
        alignContent="center"
      >
        <Grid
          container
          item
          sm={10}
          lg={8}
          xl={6}
          justifyContent="center"
          // this class name is used not only for styling but also in `scroll-to.js`

          className="main-content"
        >
          <Grid
            container
            item
            xs={12}
            md={10}
            xl={8}
            ref={null}
            direction="row"
            alignItems="flex-start"
            justifyContent="center"
            className={marginBottomX1point5}
            wrap="wrap"
          >
            <Grid item xs={12}>
              <Box>
                <Typography
                  variant="h6"
                  gutterBottom
                  align="left"
                  className={boldHeader}
                >
                  {messages.VERIFICATION}
                </Typography>
              </Box>
            </Grid>

            <Grid item xs={12}>
              <Typography paragraph align="left" variant="body2">
                {messages.CHECK_EMAIL_PART_ONE}

                {user ? obfuscateEmail(user?.email) : ''}

                {messages.CHECK_EMAIL_PART_TWO}
              </Typography>
            </Grid>

            {/* <Grid item xs={12}>
              <Typography paragraph align="left" variant="subtitle1">
                {messages.SPAM_FOLDER_CHECK}
              </Typography>
            </Grid> */}

            <Box component={Grid} pb={2}>
              <Button
                type="submit"
                onClick={handleClick}
                size="large"
                fontWeight="bold"
                variant="primary"
                showShadow={true}
                disabled={isDisabled}
                text={
                  isDisabled ? messages.EMAIL_SENT_CTA : messages.RESEND_CTA
                }
              />
            </Box>

            {apiError ? (
              <Grid item xs={12}>
                <Typography
                  paragraph
                  align="center"
                  variant="body2"
                  color="error"
                >
                  {messages.RESEND_EMAIL_ERROR}
                </Typography>
              </Grid>
            ) : null}

            <Grid item xs={12}>
              <Typography paragraph align="left" variant="subtitle1">
                {messages.HELP_TEXT_PART_ONE}
                {messages.MESSAGE_TEXT_ONE}
                <Link href={`mailto:${messages.CONTACT_CENTER_EMAIL}`}>
                  {messages.CONTACT_CENTER_EMAIL}
                </Link>
                {messages.MESSAGE_TEXT_TWO}
                <Link
                  href={`tel:${messages.CONTACT_CENTER_PHONE_NUMBER_TEL_LINK}`}
                >
                  {messages.CONTACT_CENTER_PHONE_NUMBER}
                </Link>
                {/* {messages.MEMBER_CARE_MESSAGE_TEXT_THREE} */}
              </Typography>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};
