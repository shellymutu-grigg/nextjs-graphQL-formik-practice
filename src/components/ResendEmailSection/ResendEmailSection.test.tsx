import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';
import userEvent from '@testing-library/user-event';
import { render, screen, waitFor, act } from '@utils/react-testing-utils';
import { messages } from '@utils/constants';
import { request } from '@utils/request';
import { ResendEmailSection } from './ResendEmailSection';

jest.mock('@utils/request', () => ({
  ...jest.requireActual('@utils/request'),
  request: jest.fn(),
}));

describe('Resend Email Section:', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });
  test('should match snapshot', async () => {
    mockAllIsIntersecting(true);

    const { container } = render(<ResendEmailSection />);

    await waitFor(() => expect(container).toMatchSnapshot());
  });
  test('should show enabled "Resend email" button and then display enabled "Email Sent" button after timer lapses', async () => {
    render(<ResendEmailSection />);

    const resendEmailButton = await screen.findByText(messages.RESEND_CTA);
    await waitFor(() => expect(resendEmailButton).toBeInTheDocument());

    await waitFor(() => userEvent.click(resendEmailButton));

    act(() => {
      jest.runAllTimers();
    });

    await waitFor(() =>
      expect(
        screen.getByRole('button', {
          name: messages.RESEND_CTA,
        }),
      ).toBeEnabled(),
    );
  });

  test('should disable the "resend email" button when it is clicked on by the user and enable "resend email" after timer lapses', async () => {
    render(<ResendEmailSection />);

    const resendEmailButton = await screen.findByText(messages.RESEND_CTA);
    await waitFor(() => expect(resendEmailButton).toBeInTheDocument());

    await waitFor(() => userEvent.click(resendEmailButton));
    expect(request).toHaveBeenCalledTimes(1);

    act(() => {
      jest.runAllTimers();
    });

    await waitFor(() =>
      expect(
        screen.getByRole('button', {
          name: messages.EMAIL_SENT_CTA,
        }),
      ).toBeDisabled(),
    );

    act(() => {
      jest.runAllTimers();
    });

    await waitFor(() =>
      expect(
        screen.getByRole('button', {
          name: messages.RESEND_CTA,
        }),
      ).toBeEnabled(),
    );
  });
});
