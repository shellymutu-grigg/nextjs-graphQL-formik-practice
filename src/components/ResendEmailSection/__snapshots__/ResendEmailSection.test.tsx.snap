// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resend Email Section: should match snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-root MuiGrid-container MuiGrid-align-content-xs-center MuiGrid-justify-content-xs-center"
  >
    <div
      class="MuiGrid-root main-content MuiGrid-container MuiGrid-item MuiGrid-justify-content-xs-center MuiGrid-grid-sm-10 MuiGrid-grid-lg-8 MuiGrid-grid-xl-6"
    >
      <div
        class="MuiGrid-root makeStyles-bottomMargin1point5-4 MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start MuiGrid-justify-content-xs-center MuiGrid-grid-xs-12 MuiGrid-grid-md-10 MuiGrid-grid-xl-8"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiBox-root MuiBox-root-6"
          >
            <h4
              class="MuiTypography-root makeStyles-boldHeader-3 MuiTypography-h4 MuiTypography-gutterBottom MuiTypography-alignLeft"
            >
              Verification
            </h4>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <p
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-paragraph MuiTypography-alignLeft"
          >
            If you would like us to resend verification to your email 
            
             please click the button below
          </p>
        </div>
        <div
          class="MuiGrid-root MuiBox-root MuiBox-root-7"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-shadowButton-8 MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-disableElevation"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiBox-root MuiBox-root-11"
              >
                Resend email
              </span>
            </span>
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <p
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-paragraph MuiTypography-alignLeft"
          >
            If you need help, 
            please email us at 
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="mailto:test@email.com"
            >
              test@email.com
            </a>
             or give us a call on 
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="tel:0800000000"
            >
              0800 GET HELP
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
